keps/sig-node/0028-20180906-quotas-for-ephemeral-storage.md

```diff
@@ -0,0 +1,647 @@
+  exit, and it minimizes the attack surface by not having a file that
+  can be found by an attacker.  The following pod does this; it will
+  never be caught by the present mechanism:
+
+```yaml
+apiVersion: v1
+kind: Pod
+max:
+metadata:
+  name: "diskhog"
+spec:
+  containers:
+  - name: "perl"
+    resources:
+      limits:
+        ephemeral-storage: "2048Ki"
+    image: "perl"
+    command:
+    - perl
+    - -e
+    - >
+      my $file = "/data/a/a"; open OUT, ">$file" or die "Cannot open $file: $!\n"; unlink "$file" or die "cannot unlink $file: $!\n"; my $a="0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"; foreach my $i (0..200000000) { print OUT $a; }; sleep 999999
+    volumeMounts:
+    - name: a
+      mountPath: /data/a
+  volumes:
+  - name: a
+    emptyDir: {}
+```
+* It is reactive rather than proactive.  It does not prevent a pod
+  from overshooting its limit; at best it catches it after the fact.
+  On a fast storage medium, such as NVMe, a pod may write 50 GB or
+  more of data before the housekeeping performed once per minute
+  catches up to it.  If the primary volume is the root partition, this
+  will completely fill the partition, possibly causing serious
+  problems elsewhere on the system.
+
+In many environments, these issues may not matter, but shared
+multi-tenant environments need these issues addressed.
+
+### Goals
+
+* Primary: improve performance of monitoring by using project quotas
+  in a non-enforcing way to collect information about storage
+  utilization.
+* Primary: detect storage used by pods that is concealed by deleted
+  files being held open.
+* Primary: this will not interfere with the more common user and group
+  quotas.
+* Stretch: enforce limits on per-volume storage consumption by using
+  enforced project quotas.  Each volume would be given an enforced
+  quota of the total ephemeral storage limit of the pod.
+
+### Non-Goals
+
+* Enforcing limits on total pod storage consumption by any means, such
+  that the pod would be hard restricted to the desired storage limit.
+
+## Proposal
+
+This proposal applies project quotas to emptydir volumes on qualifying
+filesystems (ext4fs and xfs with project quotas enabled).  Project
+quotas are applied by selecting an unused project ID (a 32-bit
+unsigned integer), setting a limit on space and/or inode consumption,
+and attaching the ID to one or more files.  By default (and as
+utilized herein), if a project ID is attached to a directory, it is
+inherited by any files created under that directory.
+
+If we elect to use the quota as enforcing, we impose a quota
+consistent with the desired limit.  If we elect to use it as
+non-enforcing, we impose a large quota that in practice cannot be
+exceeded (2^63-1 bytes for XFS, 2^58-1 bytes for ext4fs).
+
+### Operation Flow -- Applying a Quota
+
+* Caller (emptydir volume manager or container runtime) creates an
```

Comment:
Is enforcing quotas on container writable layers (or logs) in-scope, or out-of-scope for this proposal?  They are not listed in goals (and empty-dir quota is explicitly called out).  If it is in-scope, you should detail the changes to the CRI that are required for it. If it is not in-scope, it should be added to non-goals.  If it is out-of-scope, I am not sure if the risk about container shutdown still applies?

Reply:
I'd like it to be in scope, at least for writable layers, but who should I discuss the CRI changes with?

Reply:
@yujuhong

Reply:
This is already listed as a stretch goal.  I am clarifying that it will only be done if we can apply an enforcing quota to the writable layer.